test/include/kernels/ADCoupledConvection.h

```diff
@@ -0,0 +1,46 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+#ifndef ADCOUPLEDCONVECTION_H_
+#define ADCOUPLEDCONVECTION_H_
+
+#include "ADKernel.h"
+
+template <ComputeStage compute_stage>
+class ADCoupledConvection;
+
+template <>
+InputParameters validParams<ADCoupledConvection<RESIDUAL>>();
+template <>
+InputParameters validParams<ADCoupledConvection<JACOBIAN>>();
+
+/**
+ * Define the ADKernel for a convection operator that looks like:
+ *
+ * grad_some_var dot u'
+ *
+ */
+template <ComputeStage compute_stage>
+class ADCoupledConvection : public ADKernel<compute_stage>
+{
+public:
+  ADCoupledConvection(const InputParameters & parameters);
+
+protected:
+  virtual typename ResidualReturnType<compute_stage>::type computeQpResidual() override;
```

Comment:
Would it be worth making a typedef/macro for `ResidualReturnType<compute_stage>::type` that is just something simple... like `ADResidual` ?